request_handler_proxy.cpp

```diff
@@ -0,0 +1,187 @@
+	    response_stream >> http_version;
+
+	    unsigned int status_code;
+	    response_stream >> status_code;
+
+	    std::string status_message;
+	    std::getline(response_stream, status_message);
+	    if (!response_stream || http_version.substr(0, 5) != "HTTP/")
+	    {
+	      std::cout << "Invalid response\n";
+	      return false;
+	    }
+
+	    // Handle redirect
+	    if (status_code == 302) {
+	    	// return make_request(host, port, path, res)
+	    	// where host, port, and path are set from the redirect
+	    	// Note: there is a response parser obj as member variable of ProxyHandler to get location header
+	    	std::cout << "REDIRECT" << std::endl;
+	    	boost::asio::read_until(socket, response_buffer, "\r\n\r\n");
+	    	std::string response_string( (std::istreambuf_iterator<char>(&response_buffer)), std::istreambuf_iterator<char>() );
+	    	resp_parser.parse_response(response_string);
+	    	std::string new_loc = resp_parser.get_redirect_url();
+	    	parse_url(new_loc, host, port, path);
+	    	return make_request(host, port, path, true, res);
+	    }
+	    else if (status_code != 200)
+	    {
+	      std::cout << "Response returned with status code " << status_code << "\n";
+	      return false;
+	    }
+
+	    // Read the response headers, which are terminated by a blank line.
+	    boost::asio::read_until(socket, response_buffer, "\r\n\r\n");
+	    std::string response_string( (std::istreambuf_iterator<char>(&response_buffer)), std::istreambuf_iterator<char>() );
+
+	    //get the content type to set the correct field in the response
+	    resp_parser.parse_response(response_string);
+	    std::string content_type = resp_parser.get_content_type();
+	    if (content_type == "")
+	    	content_type = "text/plain";
+
+	    // Read until EOF, writing data to output as we go.
+	    boost::system::error_code error;
+	    // temp storage of body
+	    std::stringstream str_stream;
+	    while (boost::asio::read(socket, response_buffer,boost::asio::transfer_at_least(1), error)) {
+	      str_stream << &response_buffer;
+	    }
+	    //create a total response of the headers and message body, and isolate the message body
+	    std::string total_response = response_string + str_stream.str();
+	    std::string response_body = total_response.substr(total_response.find("\r\n\r\n")+4);
+	    // set up proxy response
+	    res->SetStatus(Response::ResponseCode::OK);
+	    res->AddHeader("Content-Type", content_type);
+	  	res->AddHeader("Content-Length", std::to_string(response_body.length()));
+	    res->SetBody(response_body);
+
+	    if (error != boost::asio::error::eof) {
+	      throw boost::system::system_error(error);
+	    }
+
+		return true;
+	}
+
+	//parse the redirect url so that it can be used for the next request
+	void ProxyHandler::parse_url(std::string url, std::string& host, std::string& port, std::string& path)
+	{
+		if (url.length() == 0)
+			return;
+		
+		if (url.find("https") != std::string::npos){
```

Comment:
yee good point, we followed the synchronous client flow then adapted it to handle redirects so there are some extra things it does that over fit spec/dont fit. I will fix this and get back to you!